### PR TITLE
Support index alias in ES|QL

### PR DESCRIPTION
**Pull request by dnhatn**
This change allows querying via aliases in ES|QL by passing the original indices to compute requests.

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequest.java [hunk 1]
```diff
@@ -32,35 +32,40 @@
 import org.elasticsearch.xpack.esql.session.EsqlConfiguration;
 
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 
-final class DataNodeRequest extends TransportRequest implements IndicesRequest {
+final class DataNodeRequest extends TransportRequest implements IndicesRequest.Replaceable {
```
Comment (n1v0lg) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequest.java:40]: As discussed on Slack, this should be a `RemoteClusterShardRequest` but that change is coming in a separate PR so not action needed.

**Review by n1v0lg — COMMENTED** [review]
The changes pertaining to local search look great to me 👍 

I still need another round with the cross-cluster changes -- also, since we are changing these would you be up for adding an alias-based test to the `RemoteClusterSecurityEsqlIT` test suite? 

Really appreciate your patience!

File: x-pack/plugin/esql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/esql/EsqlSecurityIT.java [hunk 2]
```diff
@@ -126,6 +155,49 @@ public void testAllowedIndices() throws Exception {
...
+            assertThat(
+                respMap.get("columns"),
+                equalTo(List.of(Map.of("name", "sum", "type", "double"), Map.of("name", "index", "type", "keyword")))
+            );
+            assertThat(respMap.get("values"), equalTo(List.of(List.of(72.0, "index-user2"))));
+        }
+    }
+
+    public void testAliasFilter() throws Exception {
+        for (var index : List.of("first-alias", "first-alias,index-user1", "first-*,index-")) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/esql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/esql/EsqlSecurityIT.java]: Yes, see https://github.com/elastic/elasticsearch/pull/109483/commits/ca3fed0ed405276d2f1b6d06c4e570bd96678b60.

**Comment by dnhatn** [comment]
> Adding an alias-based test to the RemoteClusterSecurityEsqlIT test suite

Sure, I added these tests in https://github.com/elastic/elasticsearch/pull/109483/commits/e4b06a681e3846990f09ae3e5ecfd8d739222b96

**Review by n1v0lg — APPROVED** [review]
LGTM. Thank you for iterating on this with me ❤️

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java [hunk 3]
```diff
@@ -87,6 +87,10 @@ public TransportSearchShardsAction(
 
     @Override
     protected void doExecute(Task task, SearchShardsRequest searchShardsRequest, ActionListener<SearchShardsResponse> listener) {
+        searchShards(task, searchShardsRequest, listener);
+    }
+
+    public void searchShards(Task task, SearchShardsRequest searchShardsRequest, ActionListener<SearchShardsResponse> listener) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchShardsAction.java:97]: ++, I added it in https://github.com/elastic/elasticsearch/pull/109483/commits/00ecac557c2b83a00190451de9096885bffdccf9

**Review by nik9000 — APPROVED** [review]

File: x-pack/plugin/esql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/esql/EsqlSecurityIT.java [hunk 4]
```diff
@@ -124,6 +126,20 @@ public void testAllowedIndices() throws Exception {
             assertThat(respMap.get("columns"), equalTo(List.of(Map.of("name", "sum", "type", "double"))));
             assertThat(respMap.get("values"), equalTo(List.of(List.of(72.0))));
         }
+        for (var index : List.of("my-alias-to-index-user2", "my-alias-to-index-user2,index-user2", "my-alias-*", "my-alias-*,index*")) {
+            Response resp = runESQLCommand("alias_user2", "from " + index + " | stats sum=sum(value)");
+            assertOK(resp);
+            Map<String, Object> respMap = entityAsMap(resp);
+            assertThat(respMap.get("columns"), equalTo(List.of(Map.of("name", "sum", "type", "double"))));
+            assertThat(respMap.get("values"), equalTo(List.of(List.of(72.0))));
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/esql/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/esql/EsqlSecurityIT.java]: Yes, I adjusted it in https://github.com/elastic/elasticsearch/pull/109483/commits/3db3cff9581c39ff5c232c1e82e63f76d99e15d3.

**Comment by dnhatn** [comment]
@n1v0lg @nik9000 Thank you for your feedback and review :)

**Pull request merged**